cpp/include/cuml/fil/forest_model.hpp

```diff
@@ -298,8 +298,12 @@ struct forest_model {
                std::optional<index_type> specified_chunk_size = std::nullopt)
   {
     // TODO(wphicks): Make sure buffer lands on same device as model
-    auto out_buffer = raft_proto::buffer{output, num_rows * num_outputs(), out_mem_type};
-    auto in_buffer  = raft_proto::buffer{input, num_rows * num_features(), in_mem_type};
+    int current_device_id;
+    raft_proto::cuda_check(cudaGetDevice(&current_device_id));
```

Comment:
I wonder how much runtime overhead this will create at inference time.

One way to avoid the call to `cudaGetDevice` is to assume that inference buffers reside on the same device as the model.

Reply:
> One way to avoid the call to `cudaGetDevice` is to assume that inference buffers reside on the same device as the model.

It seems that the TODO comment implies to use that approach?

Reply:
The comment says `Make sure buffer lands on same device as model`. I interpret this as, make sure that the buffer and the model land on the same device, by making device-to-device copy as necessary. This is what this PR currently does.

Reply:
We have two options:

* Current approach: query current device using `cudaGetDevice`.
  - Pro: More robust against user error (setting incorrect device)
  - Con: Extra call to `cudaGetDevice`, with possible performance overhead; implicit copy made if the model and the buffer are in different devices
* Alternative approach: Don't query current device using `cudaGetDevice`; assume that buffers would reside on the same device as the model.
  - Pro: Eliminates the `cudaGetDevice` call
  - Con: If the user puts the model and the buffer in different devices, the program will crash.